[PyTorch] TensorIteratorBase structured kernels must set dtype

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55875 [PyTorch] Make as_strided_ use_const_ref_for_mutable_tensors
* #55691 [PyTorch] Migrate add operators to borrow in TensorIteratorBase
* #55690 [PyTorch] Support borrowing in/out Tensors in TensorIterator
* **#55689 [PyTorch] TensorIteratorBase structured kernels must set dtype**
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned

Upcoming changes don't allow structured kernels to rely on
the path in `TensorIteratorBase::set_output` that notices that output
tensors were undefined and sets their current_type if so. Instead,
they must cooperate with `TensorIteratorBase` to set up the dtype.

The specific upcoming changes are that TensorIteratorBase's
OperandState::tensor member may be a borrow from the `outputs_` array
and thus will become defined when the structured kernel subclass sets
up the `outputs_` array, missing the dtype initialization. I also plan
to remove the `outputs_` array entirely in favor of the existing
`OperandState::tensor` member.

Differential Revision: [D27680973](https://our.internmc.facebook.com/intern/diff/D27680973/)

swolchok committed:
[PyTorch] TensorIteratorBase structured kernels must set dtype

Upcoming changes don't allow structured kernels to rely on
the path in `TensorIteratorBase::set_output` that notices that output
tensors were undefined and sets their current_type if so. Instead,
they must cooperate with `TensorIteratorBase` to set up the dtype.

The specific upcoming changes are that TensorIteratorBase's
OperandState::tensor member may be a borrow from the `outputs_` array
and thus will become defined when the structured kernel subclass sets
up the `outputs_` array, missing the dtype initialization. I also plan
to remove the `outputs_` array entirely in favor of the existing
`OperandState::tensor` member.

Differential Revision: [D27680973](https://our.internmc.facebook.com/intern/diff/D27680973/)

[ghstack-poisoned]

ezyang commented:
Just to verify, #55690 is specifically the PR that breaks this, right?

ezyang commented:
Hmm. I don't want to do this. Let me see if I can propose an alternative.

One reason I don't like this is, if I understand correctly, this segment of `TensorIteratorBase::set_output` is now dead:

```
  if (!op.tensor.defined()) {
    // this is dead
    op.tensor = t;
    op.current_dtype = op.target_dtype;
  }
```

So (although this couples the change to the next PR), I'd like it if we deleted this branch unconditionally so people don't think that it is possible for op.tensor to be undefined.

The next problem is that the cached `current_dtype` entry is inaccurate. But there is a simple invariant describing what this should be: `op.tensor.dtype()`. (Note that the caching might not even be necessary now, since we've devirtualized a bunch of Tensor methods). But lets say we still need this to be up to date. We ought to then simply just do an unconditional assignment that refreshes this with the correct value: `op.current_dtype = op.tensor.dtype()` (TBH, I'm not sure why the existing code doesn't do that, we could probably test out that this is sound by ONLY making this change in a PR). Then no more codegen changes are necessary. WDYT?

swolchok commented:
Sounds good to me. The dtype syncing is causing problems with my still-in-progress change to remove the outputs array from TensorIterator-based structured kernels, so fixing it would help anyway.

swolchok committed:
rebase on "[PyTorch] TensorIteratorBase structured kernels must set dtype"


Upcoming changes don't allow structured kernels to rely on
the path in `TensorIteratorBase::set_output` that notices that output
tensors were undefined and sets their current_type if so. Instead,
they must cooperate with `TensorIteratorBase` to set up the dtype.

The specific upcoming changes are that TensorIteratorBase's
OperandState::tensor member may be a borrow from the `outputs_` array
and thus will become defined when the structured kernel subclass sets
up the `outputs_` array, missing the dtype initialization. I also plan
to remove the `outputs_` array entirely in favor of the existing
`OperandState::tensor` member.

Differential Revision: [D27680973](https://our.internmc.facebook.com/intern/diff/D27680973/)

[ghstack-poisoned]

swolchok committed:
fix newly-added resize_as_sparse_ on "[PyTorch] TensorIteratorBase structured kernels must set dtype"


Upcoming changes don't allow structured kernels to rely on
the path in `TensorIteratorBase::set_output` that notices that output
tensors were undefined and sets their current_type if so. Instead,
they must cooperate with `TensorIteratorBase` to set up the dtype.

The specific upcoming changes are that TensorIteratorBase's
OperandState::tensor member may be a borrow from the `outputs_` array
and thus will become defined when the structured kernel subclass sets
up the `outputs_` array, missing the dtype initialization. I also plan
to remove the `outputs_` array entirely in favor of the existing
`OperandState::tensor` member.

Differential Revision: [D27680973](https://our.internmc.facebook.com/intern/diff/D27680973/)

[ghstack-poisoned]

swolchok commented:
> an unconditional assignment that refreshes this with the correct value: op.current_dtype = op.tensor.dtype()

I have three problems with this:
1) It is unnecessary memory traffic when the cache should be up to date.
2) I am not sure that it is strictly a cache. allocate_or_resize_outputs() sets `op.current_dtype = op.target_dtype`.
3) why do we cache current_dtype at all? op.tensor.dtype() is already a plain old load; we're just "caching" one pointer hop closer, but we do a lot of extra memory operations to maintain that cache.

swolchok commented:
> Just to verify, #55690 is specifically the PR that breaks this, right?

it breaks once you try to use the borrowing support added in that PR, yes.

ezyang commented:
> why do we cache current_dtype at all? op.tensor.dtype() is already a plain old load; we're just "caching" one pointer hop closer, but we do a lot of extra memory operations to maintain that cache.

I am 100% OK with dumping the cache. I think it's a holdover from when it was virtualized / dtypes were expensive.

swolchok commented:
> But lets say we still need this to be up to date. We ought to then simply just do an unconditional assignment that refreshes this with the correct value: op.current_dtype = op.tensor.dtype()

This isn't quite right. If I do that in  #55690  and also remove the `if (!op.tensor.defined())` block you say is dead above, I start failing assertions. I suspect that the root cause is that that block is *not* dead for structured kernels that aren't migrated to borrowing.

swolchok commented:
>  that block is not dead for structured kernels that aren't migrated to borrowing.

It may die in the future (like if I'm successful at removing the `outputs_` array for TensorIteratorBase structured kernels), but it won't die just because of borrowing support.

bhosmer commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -224,7 +224,7 @@ def gen_class_set_output(self, k: SchemaKind, parent_class: str, generate_super:
 """
 
-    def gen_class_set_output_body(self, k: SchemaKind) -> str:
+    def gen_class_set_output_body(self, k: SchemaKind, parent_class: str) -> str:
```
Is `parent_class` used in here?

bhosmer reviewed (commented):
This LGTM but I'll leave it to @ezyang to approve, as I'm not sure where the thread wound up w.r.t. his comfort level with this.

ezyang commented:
I'd still like to find another way to do this, I'm OK to approve this to unblock the stack. I don't remember what the status of this PR set is.

swolchok commented:
we don't need this, abandoning
